barcode_utils.py

```diff
@@ -0,0 +1,166 @@
+from barcode import *
+
+import base64
+import struct
+import skip32
+import code128
+import os
+
+
+# TODO: currently broken, need to rewrite in the context of uber
+def generate_barcode_csv(filename):
+    badge_types = self.badge_types['badge_types']
+
+    # TODO: modify so we read the config directly from uber.
+    lines = []
+    for badge_type, ranges in badge_types.items():
+        range_start = int(ranges['range_start'])
+        range_end = int(ranges['range_end'])
+        lines = lines + self.generate_barcode_nums(range_start, range_end)
+
+    f = open(filename,'w')
+
+    for line in lines:
+        f.write(line + os.linesep)
+
+    f.close()
+
+
+# TODO: currently broken, need to rewrite in the context of uber
+def generate_barcode_nums(range_start, range_end):
+    generated_lines = []
+    seen_barcodes = []
+    for badge_num in range(range_start, range_end+1):
+        barcode_num = generate_barcode_from_badge_num(
+            badge_num=int(badge_num),
+            event_id=self.event_id,
+            salt=self.salt,
+            key=self.secret_key
+        )
+
+        line = "{badge_num},{barcode_num}".format(
+            badge_num=badge_num,
+            barcode_num=barcode_num,
+        )
+        generated_lines.append(line)
+
+        # ensure that we haven't seen this value before
+        if barcode_num in seen_barcodes:
+            raise ValueError('COLLISION: generated a badge# that\'s already been seen')
+        else:
+            seen_barcodes.append(barcode_num)
+
+    return generated_lines
+
+
+def generate_barcode_from_badge_num(badge_num, event_id=None, salt=None, key=None):
+    event_id = config['secret']['barcode_event_id'] if not event_id else event_id
+    salt = config['secret']['barcode_salt'] if not salt else salt
+    key = bytes(config['secret']['barcode_key'],'ascii') if not key else key
+
+    # packed data going to be encrypted is:
+    # byte 1 - 8bit event ID, usually 1 char
+    # byte 2,3,4 - 24bit badge number
+
+    salted_val = badge_num + (0 if not salt else salt)
+
+    if salted_val > 0xFFFFFF:
+        raise ValueError("badge_number is too high " + str(badge_num))
+
+    data_to_encrypt = struct.pack('>BI', event_id, salted_val)
+
+    # remove the highest byte in that integer (2nd byte)
+    data_to_encrypt = bytearray([data_to_encrypt[0], data_to_encrypt[2], data_to_encrypt[3], data_to_encrypt[4]])
+
```

Comment:
Ehhh, while it's technically true that Python can be run with a command line flag that turns off all assertions, 1) I've never actually seen anyone run Python that way, and 2) we ourselves do not run it that way.  My feeling is it's okay to use asserts even for data validation for those reasons.

Some people disagree, obviously.  For example, ``undefined`` is just a global constant in Javascript and isn't a language builtin.  For this reason, some people say that you should never refer to ``undefined`` directly, e.g.

```javascript
if (foo === undefined) { }  // may break if someone overrides undefined

if (typeof foo === 'undefined') { }  // will always work
```

My response to this is basically, "Oh c'mon, seriously?  Don't override ``undefined``.  Just don't ever do it.  If you install a third party module which overrides ``undefined`` then uninstall it and BURN IT WITH FIRE because NO ONE SHOULD EVER BE DOING THAT!!!!!"

I feel basically the same way about ``assert`` in Python.  Not a huge deal, but if someone says, "We should write extra code in case someone sets a flag they probably shouldn't be setting" then my response is "don't set that flag".  Just my 2 cents.